File: android2_lesson04/src/java/java/com/example/ama/android2_lesson04/ui/viewer/PVFragmentAsynkTask.java
package com.example.ama.android2_lesson04.ui.viewer;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import com.example.ama.android2_lesson04.R;
import com.example.ama.android2_lesson04.background.asynk.LoadingPicturesAsynkTask;
import com.example.ama.android2_lesson04.ui.viewer.base.BasePictureViewerFragment;

public class PVFragmentAsynkTask extends BasePictureViewerFragment {

    private LoadingPicturesAsynkTask asynkTask;

    public static PVFragmentAsynkTask newInstance() {
        return new PVFragmentAsynkTask();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        setRetainInstance(true);
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (asynkTask != null) {
            asynkTask.setListener(getListener());
        }
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    protected void onLoadImagesClick() {
        if (asynkTask == null) {
            asynkTask = new LoadingPicturesAsynkTask(getListener());
            asynkTask.execute(getResources().getStringArray(R.array.image_urls));
        }
    }

    private LoadingPicturesAsynkTask.OnLoadingListener getListener() {
        return new LoadingPicturesAsynkTask.OnLoadingListener() {
            @Override
            public void onUpdate(Bitmap bitmap) {
                onUpdateLoading(bitmap);
            }

            @Override
            public void onFinish() {
                onFinishLoading();
                asynkTask = null;
            }
        };
    }

    @Override
    public void onDestroyView() {
        if (asynkTask != null) {
            asynkTask.removeListener();
        }
        super.onDestroyView();
    }
}

File: android2_lesson04/src/kotlin/java/com/example/ama/android2_lesson04/ui/viewer/PVFragmentAsynkTask.kt
package com.example.ama.android2_lesson04.ui.viewer

import android.graphics.Bitmap
import android.os.Bundle
import android.view.View
import com.example.ama.android2_lesson04.R
import com.example.ama.android2_lesson04.background.asynktask.LoadPicturesAsynkTask
import com.example.ama.android2_lesson04.ui.viewer.base.BasePictureViewerFragment

class PVFragmentAsynkTask : BasePictureViewerFragment() {

    companion object {
        fun newInstance() = PVFragmentAsynkTask()
    }

    private var asynkTask: LoadPicturesAsynkTask? = null
    private val updateListener = { bitmap: Bitmap ->
        onUpdateLoading(bitmap)
    }
    private val finishListener = {
        onFinishLoading()
        asynkTask = null
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        retainInstance = true
        super.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        asynkTask?.setListeners(finishListener, updateListener)
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onLoadImagesClick() {
        if (asynkTask == null) {
            asynkTask = LoadPicturesAsynkTask(finishListener, updateListener)
            asynkTask?.execute(*resources.getStringArray(R.array.image_urls))
        }
    }

    override fun onDestroyView() {
        asynkTask?.removeListeners()
        super.onDestroyView()
    }
}